### PR TITLE
New lint: zst_offset

### LINKED ISSUE #4813: New lint: NOP pointer arithmetic
(I'm not sure if clippy has enough information for this, but it would've caught a big error I just fixed.)

Clippy should lint on the use of `<*[const|mut] T>::offset`, `wrapping_offset`, `add`, `wrapping_add`, `sub`, and `wrapping_sub` where `T` has size 0. This is a NOP and likely unintended.

### PR DESCRIPTION
**Pull request by llogiq**
This fixes #4813 

changelog: add `zst_offset` lint

### CODE REVIEW

**Comment by JohnTitor** [comment]
Seems occurred an ICE on dogfooding

File: clippy_lints/src/methods/mod.rs [hunk 1]
```diff
@@ -3063,3 +3084,15 @@ fn contains_return(expr: &hir::Expr) -> bool {
     visitor.visit_expr(expr);
     visitor.found
 }
+
+fn check_pointer_offset(cx: &LateContext<'_, '_>, expr: &hir::Expr, args: &[hir::Expr]) {
+    if_chain! {
+        if args.len() == 2;
+        if let ty::RawPtr(ty::TypeAndMut { ref ty, .. }) = cx.tables.expr_ty(&args[0]).kind;
+        if let Ok(layout) = cx.tcx.layout_of(ty::ParamEnv::empty().and(ty));
```
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs]: `TyLayout` has a `is_zst` method. 

Also this panics in dogfood, becacuse `normalize_erasing_regions` fails.

Comment (flip1995) [reply clippy_lints/src/methods/mod.rs]: There is also [`rustc_lint::types::is_zst`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint/types/fn.is_zst.html), which requires a `def_id` for building the `ParamEnv`, before calling `layout_of`:

```rust
fn is_zst<'tcx>(tcx: TyCtxt<'tcx>, did: DefId, ty: Ty<'tcx>) -> bool {
    tcx.layout_of(tcx.param_env(did).and(ty)).map(|layout| layout.is_zst()).unwrap_or(false)
}
```

**Comment by llogiq** [comment]
@flip1995 r?

File: clippy_lints/src/methods/mod.rs [hunk 2]
```diff
@@ -1065,6 +1065,23 @@ declare_clippy_lint! {
...
+    /// **What it does:** Checks for `offset(_)`, `wrapping_`{`add`, `sub`}, etc. on raw pointers to
+    /// zero-sized types
+    ///
+    /// **Why is this bad?** This is a no-op, and likely unintended
+    ///
+    /// **Known problems:** None
+    ///
+    /// **Example:**
+    /// ```ignore
+    /// unsafe { (&() as *const ()).offest(1) };
```
Comment (CAD97) [inline clippy_lints/src/methods/mod.rs]: s/offest/offset

Comment (llogiq) [reply clippy_lints/src/methods/mod.rs]: Will do a followup.

**Pull request merged**